### ISSUE TITLE
Proper way to retrieve historical hashtag data

### ISSUE THREAD

**Issue by sp1thas** [instaloader/instaloader#469]
I'm trying to download historical data for a list of hashtag. The list of hashtags contains highly populated and popular hashtags (I.e [`nike`](https://www.instagram.com/explore/tags/nike/)) with millions of posts. I ideally want to download images and meta data of posts for the last 12 months.

My first thought was to split date frame in smaller parts and perform download process multiple times using a smaller date window (1 month) every time.

After few tries I came up with some difficulties. During my attempts I used `instaloader` as a python module.

### 1. posts are not fully sorted based on publish date
While iterating over `L.get_hashtag_posts(hashtag)` some posts have totally different post date:
```text
>>> for post in L.get_hashtag_posts('rebook'):
>>>     print(post.date)
... [...]
... 2019-10-18 02:58:45
... 2019-10-18 02:58:02
... 2019-10-16 16:57:16
... 2019-10-16 13:52:37
... 2019-09-06 12:27:16 # ??
... 2018-09-11 04:01:39 # ??
... 2019-10-18 02:57:26
... 2019-10-18 02:57:08
... 2019-10-18 02:56:26
... 2019-10-18 02:40:06
... 2019-10-18 02:39:28
... 2019-10-18 02:38:42
... 2019-10-18 01:59:21
... 2019-10-18 01:51:27
... 2019-10-18 01:50:59
... 2019-10-18 01:46:48
... 2019-10-18 01:45:16
... [...]
```
As a result, [this](https://instaloader.github.io/codesnippets.html#download-posts-in-a-specific-period) example is not working.

### 2. for loop infinitely over  `L.get_hashtag_posts(hashtag)` and filter manually
Due to the fact that posts are not fully sorted, I decided to iterate over `L.get_hashtag_posts('rebook')` and check every post's date.
After few thousands of posts I got the following error:
```
JSON Query to explore/tags/rebook/: 429 Too Many Requests [retrying; skip with ^C]
JSON Query to explore/tags/rebook/: 429 Too Many Requests [retrying; skip with ^C]
Traceback (most recent call last):
  File "/home/simakis/.p3env/lib/python3.6/site-packages/instaloader/instaloadercontext.py", line 371, in get_json
    raise TooManyRequestsException("429 Too Many Requests")
instaloader.exceptions.TooManyRequestsException: 429 Too Many Requests

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/simakis/.p3env/lib/python3.6/site-packages/instaloader/instaloadercontext.py", line 371, in get_json
    raise TooManyRequestsException("429 Too Many Requests")
instaloader.exceptions.TooManyRequestsException: 429 Too Many Requests

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/simakis/.p3env/lib/python3.6/site-packages/instaloader/instaloadercontext.py", line 371, in get_json
    raise TooManyRequestsException("429 Too Many Requests")
instaloader.exceptions.TooManyRequestsException: 429 Too Many Requests

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "harv.py", line 202, in <module>
    hv.download()
  File "harv.py", line 166, in download
    for post in posts:
  File "/home/simakis/.p3env/lib/python3.6/site-packages/instaloader/instaloader.py", line 833, in get_hashtag_posts
    params)['graphql']['hashtag']['edge_hashtag_to_media']
  File "/home/simakis/.p3env/lib/python3.6/site-packages/instaloader/instaloadercontext.py", line 399, in get_json
    return self.get_json(path=path, params=params, host=host, session=sess, _attempt=_attempt + 1)
  File "/home/simakis/.p3env/lib/python3.6/site-packages/instaloader/instaloadercontext.py", line 399, in get_json
    return self.get_json(path=path, params=params, host=host, session=sess, _attempt=_attempt + 1)
  File "/home/simakis/.p3env/lib/python3.6/site-packages/instaloader/instaloadercontext.py", line 392, in get_json
    raise ConnectionException(error_string) from err
instaloader.exceptions.ConnectionException: JSON Query to explore/tags/rebook/: 429 Too Many Requests
```
I would like to ask you if there are any best practices for this kind of task and if it is feasible to download historical data from highly populated hashtags.

**Comment by aandergr** [comment]
Related to #464 in that it refers to not yet implemented rate control for certain endpoints.

**Comment by aandergr** [comment]
Should be fixed with de585becdb0be5082ded70729888792bf9889d1b (to be released with 4.3). To already get a version with that fix included, do

```
pip install --upgrade git+https://github.com/instaloader/instaloader
```

Please comment whether this fixes the issue.
